### ISSUE TITLE
[jsx-no-useless-fragment] Option to allow "<></>"

### ISSUE THREAD

**Issue by SrBrahma** [yannickcr/eslint-plugin-react#3194]
For Typescript, when we don't want our component to render at all, we do `return <></>` instead of just `return`, as the latter would throw a type error, as it's expected to return JSX.Element etc.

`jsx-no-useless-fragment` allow us to have a `<>{}</>` for that case, but it's just ugly and it's extra characters to type.

I suggest having an `allowEmpty` option, that would make the rule don't complain about `<></>`.

**Comment by ljharb** [comment]
The idiomatic react thing to do here has always been `return null`; creating a throwaway fragment serves no purpose.

**Comment by SrBrahma** [comment]
It does serve a purpose.

It's a common pattern in TS for the functions to return JSX.Element type, and it doesn't allow returning `null`. (https://github.com/typescript-cheatsheets/react#function-components)

![image](https://user-images.githubusercontent.com/28177550/151741507-fe91789c-2c8d-4834-8d98-dc240e6e1c8e.png)

**Comment by wimbarelds** [comment]
Even if something is typed as `React.ReactNode`, components may checking for truthiness of a prop. An empty fragment is truthy whereas `null` is falsy. The difference between an empty fragment and null _can_ communicate different intent in a clear/reasonable way.

When enabling this rule and 'fixing' violations I ran into the problem where replacing empty fragments with null was causing the application's behavior to change in an unwanted way. Could 'allowEmpty' be an option?

**Comment by ljharb** [comment]
@wimbarelds can you provide a concrete example of when it's reasonable to check for truthiness/falsiness, and when you want something truthy that renders nothing?

**Comment by wimbarelds** [comment]
> @wimbarelds can you provide a concrete example of when it's reasonable to check for truthiness/falsiness, and when you want something truthy that renders nothing?

En ErrorBoundary component's fallback prop, with something akin to:
```
return fallback ?? <Alert msg={`Error occured rendering ${componentName}`} />
```

There may be cases where you just want to just log that error and not show it to the user.

Yes code could be changed to do something like `'fallback' in props` or distinguishing between null and undefined; but to me that would feel like it's leaning a lot more towards writing code to appease your linter than for quality purposes.

**Comment by ljharb** [comment]
I would say that producing an object and rendering it rather than writing a more precise check is not about appeasing the linter - in that case, i'd suggest putting that Alert in a `defaultProps` rather than recreating it every render. That way, you get a performance boost, a clarity boost, AND you appease the linter.

**Comment by wimbarelds** [comment]
> I would say that producing an object and rendering

To clarify, by this you're referring to `<></>`?

Beyond that.
* I'm not quite following how Alert taking a child helps instead of a string prop changes the performance significantly. Honestly this is mostly an idle curiousity- the ErrorBoundary example is similar to an actual issue but its not an exact copy.
* It feels... misplaced, to put as much emphasis or care at the render performance of an ErrorBoundary to the point that a single empty fragment would be sufficiently meaningful. This is a portion of code that should not be reached more than sporadically. The readability and bundle size are bigger concerns.

With regards to performance though (because it is kind of an interesting brain teaser 🤷‍♂️), the instantiation of the (Alert) component happening in the default props would/could only be better if the errorboundary is actually rendering the fallback (more than once) right?

Syntactically / organizationally, I don't find default props to be particularly appealing when it comes to longer values (or JSX values in general).

The current workaround is `fallback={<>{null}</>}`, and that null expression is definitely just for the benefit of eslint.

**Comment by ljharb** [comment]
I agree adding that null is just for eslint. I'm just not convinced that it makes sense to provide an empty fragment in N consumers of the ErrorBoundary, versus allowing, say, an empty string - with your `??` check, wouldn't `fallback=""` work fine?

**Comment by wimbarelds** [comment]
In my eyes, an empty fragment as a prop feels like a much more explicit expression of intent than an empty string as prop (which would also require a minor change to the interface, but 🤷‍♂️).

I think the question should be more one of, are there reasonable situations in which someone might wish to use empty fragments without it being problematic / an error? There is a valid point of performance, but I also think there's a valiid argument to be made for semantics / 'what does this code communicate'- in which I think there is at least somewhat of a fair argument to be made for empty fragments.

If that's not a compelling enough argument, I would request the documentation page communicating something along the lines of "hey, `null` is usually a good substitute for an empty fragment; but be aware that `<></>` is truthy and null is not.

**Comment by ljharb** [comment]
hmm. Note that I'm not saying I can't be convinced, more that I'd need multiple compelling use cases to convince me that it's reasonable (and that there's likely to be more use cases out there).

As far as the docs, absolutely I'd be happy to review a PR that talks about the various ways to "render nothing", and their tradeoffs.
